warrior/WarriorCore/Classes/kw_driver_class.py

```diff
@@ -158,33 +159,66 @@ def get_mandatory_arguments(self):
             args.remove('self')
         return args
 
+    def get_credential_value(self, arg, system):
+        """get the value of arg in data file corresponding to system
+        """
+        datafile = self.data_repository['wt_datafile']
+        var = arg
+        if not hasattr(self, 'tag_dict'):
+            self.tag_dict = data_Utils.get_credentials(datafile, system)
+            arg = arg.strip()
+        if isinstance(arg, basestring) and arg.startswith("wtag"):
+            var = arg.split("=")[1].strip()
+        if var in self.tag_dict:
+            value = self.tag_dict[var]
+            # substitute environment/datarepo variables in the value and return
+            if isinstance(value, (basestring, list, dict)):
+                return data_Utils.substitute_var_patterns(value)
+            else:
+                return value
+        else:
+            return None
```

Comment:
1 - added inline comments explaining the need. Basically, this is the place where we are converting wtag var to the actual value if wtag var is used
2 - this is for the case where the mandatory var is not menioned in the args but it need to be picked up from wtags as per our wtags design doc.
3 - yes, it makes it easier for me to access it. Please let me know, if you have any other suggestions
  a. do you want to move the initialization part to the constructor? But this requires accessing system_name arg and calling get_credentials function. 
  b. from point a, if the initialization is moved to the constructor part, then the destruction can be done in destructor part. Basically, need a mechanism to clean up tag_dict after processing both mandatory and optional args. Doing it in get_values_for_mandatory_args would make it unavailable in get_values_for_optional_args function.
   c. yes, but it needs to be passed around in atleast these three functions - get_values_for_mandatory_args, get_values_for_optional_args and get_credential_value

4 - I am checking if the arg is starting with wtag and then only splitting on "=". What advantage would I get if I split on "wtag="? Atleast with the other approach, I can handle white spaces around the operators which won't be possible in this approach.

Reply:
@satyakrish
#1: 
 - wtag  feature would not work if system name is not specified
 - if it is not specified in first place the wtag feature can not be used for the arguments before the system name argument is specified
 - if code is added to check for system name in the args list and then process each argument, then this adds unnecessary cycles for each keyword
#2: 
targ = arg_kv[args] if args in arg_kv else args
if args is arg_int and the user has passed 'wtag=intvalue' for this arg then targ would be assigned 'wtag=intvalue' otherwise it would be arg_int and if this arg_int is defined in tag_dict, that should be picked up as per our design document, which is what this get_credential_value() does. 
#3: This is indeed advisable practice. Please refer https://stackoverflow.com/questions/45486395/what-happens-when-you-initialise-instance-variables-outside-of-init."As a dynamic language, Python behaves in exactly the same way wherever you set an attribute, and __init__ is not in any way privileged when it comes to creating attributes."

Reply:
Can we discuss this in a call? Please let me know a convenient time when we can discuss this.

Reply:
@navvrnd  yes pls schedule for a call from outlook during my working hours.